OMEdit/OMEdit/OMEditGUI/Modeling/ModelWidgetContainer.cpp

```diff
@@ -1503,8 +1500,28 @@ void GraphicsView::addItem(QGraphicsItem *pGraphicsItem)
  */
 void GraphicsView::removeItem(QGraphicsItem *pGraphicsItem)
 {
-  if (scene()->items().contains(pGraphicsItem)) {
+  if (mAllItems.contains(pGraphicsItem)) {
+    mAllItems.remove(pGraphicsItem);
+    scene()->removeItem(pGraphicsItem);
+  }
+}
+
+/*!
+ * \brief GraphicsView::removeItem
+ * Removes the QGraphicsItem from GraphicsView and either deletes it immediately
+ * or postpones it if mToBeCleared is set.
+ * \param pGraphicsItem item to remove (should be an instance of Component if doEmitDeleted is true)
+ * \param doEmitDeleted whether emit deleted() between removal and deletion or not
+ */
+void GraphicsView::removeItemAndDelete(QGraphicsItem *pGraphicsItem, bool doEmitDeleted)
+{
+  if (!mToBeCleared && mAllItems.contains(pGraphicsItem)) {
```

Comment:
And here we don't check for `mToBeCleared` since we really want to remove the item and call `emitDeleted`. If `mToBeCleared` then we miss the `emitDeleted` and some views might end up wrong.

Reply:
Unfortunately, simply reverting to the old behavior when `doEmitDeleted` is set will ruin the performance again.

Are reference components always in the same view? If so, then we can probably ignore `referenceComponentDeleted()` but should care about `handleIconUpdated()`? If so, then it can be be unconditionally called right after clear.